src/ShaderGen/ShaderMethodVisitor.cs

```diff
@@ -269,6 +282,55 @@ public override string VisitInvocationExpression(InvocationExpressionSyntax node
             }
         }
 
+        /// <summary>
+        /// Given an expression containing a reference to a texture object, this method
+        /// finds the corresponding class instance field declaration(s) for those texture(s).
+        /// The expression may be a direct reference to the field, or it may refer to a
+        /// method parameter, in which case we need to find all texture fields that are passed
+        /// as arguments to that method.
+        /// </summary>
+        /// <returns></returns>
+        private void FindTextureFieldsRecursive(ExpressionSyntax expression, List<IFieldSymbol> fieldSymbols, List<IParameterSymbol> parameterSymbols)
+        {
+            SemanticModel semanticModel = _compilation.GetSemanticModel(expression.SyntaxTree);
+            ISymbol textureSymbol = semanticModel.GetSymbolInfo(expression).Symbol;
+            switch (textureSymbol.Kind)
+            {
+                case SymbolKind.Field:
+                    fieldSymbols.Add((IFieldSymbol) textureSymbol);
+                    break;
+
+                case SymbolKind.Parameter:
+                    IParameterSymbol parameterSymbol = (IParameterSymbol) textureSymbol;
+                    parameterSymbols.Add(parameterSymbol);
+                    int parameterIndex = parameterSymbol.Ordinal;
+                    MethodDeclarationSyntax methodDeclaration = expression.FirstAncestorOrSelf<MethodDeclarationSyntax>();
+                    ISymbol thisMethodSymbol = semanticModel.GetDeclaredSymbol(methodDeclaration);
+                    foreach (var syntaxTree in _compilation.SyntaxTrees)
+                    {
+                        SemanticModel otherSemanticModel = _compilation.GetSemanticModel(syntaxTree);
+
+                        List<InvocationExpressionSyntax> invocationsOfThisMethod = syntaxTree.GetRoot()
+                            .DescendantNodes()
+                            .OfType<InvocationExpressionSyntax>()
+                            .Where(x => otherSemanticModel.GetSymbolInfo(x).Symbol.Equals(thisMethodSymbol))
+                            .ToList();
+
+                        foreach (InvocationExpressionSyntax invocation in invocationsOfThisMethod)
+                        {
+                            FindTextureFieldsRecursive(
+                                invocation.ArgumentList.Arguments[parameterIndex].Expression, 
+                                fieldSymbols,
+                                parameterSymbols);
+                        }
+                    }
+                    break;
+
+                default:
+                    throw new NotImplementedException();
```

Comment:
Let's change the `NotImplementedException` instances to ShaderGenerationException with at least some kind of message before we merge it in.